tests/main_test.go

```diff
@@ -0,0 +1,258 @@
+//go:build integration
+
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tests
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+// TestMain builds the pebble binary before running the integration tests.
+func TestMain(m *testing.M) {
+	goBuild := exec.Command("go", "build", "-o", "../pebble", "../cmd/pebble")
+	if err := goBuild.Run(); err != nil {
+		fmt.Println("Cannot build pebble binary:", err)
+		os.Exit(1)
+	}
+
+	exitCode := m.Run()
+	os.Exit(exitCode)
+}
+
+func createLayer(t *testing.T, pebbleDir string, layerFileName string, layerYAML string) {
+	t.Helper()
+
+	layersDir := filepath.Join(pebbleDir, "layers")
+	err := os.MkdirAll(layersDir, 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers directory: pipe: %v", err)
+	}
+
+	layerPath := filepath.Join(layersDir, layerFileName)
+	err = os.WriteFile(layerPath, []byte(layerYAML), 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func createIdentitiesFile(t *testing.T, pebbleDir string, identitiesFileName string, identitiesYAML string) {
+	t.Helper()
+
+	identitiesPath := filepath.Join(pebbleDir, identitiesFileName)
+	if err := os.WriteFile(identitiesPath, []byte(identitiesYAML), 0o755); err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func pebbleRun(t *testing.T, pebbleDir string, args ...string) <-chan servicelog.Entry {
+	t.Helper()
+
+	logsCh := make(chan servicelog.Entry)
+
+	cmd := exec.Command("../pebble", append([]string{"run"}, args...)...)
+	cmd.Env = append(os.Environ(), "PEBBLE="+pebbleDir)
+
+	t.Cleanup(func() {
```

Comment:
I think this `Cleanup` call should be moved to after the `cmd.Start` call, in case the test fails before then, we don't have anything to clean up as the command hasn't started yet.